azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,122 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import threading
+
+import traceback
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.future import ustr, PriorityQueue
+from azurelinuxagent.common.interfaces import ThreadHandlerInterface
+
+
+def get_telemetry_service_handler(protocol_util):
+    return TelemetryServiceHandler(protocol_util)
+
+
+class TelemetryServiceHandler(ThreadHandlerInterface):
+    """
+    This Handler takes care of sending all telemetry out of the agent to Wireserver. It sends out data as soon as
+    there's any data available in the queue to send.
+    """
+
+    _THREAD_NAME = "TelemetryServiceHandler"
+    _MAX_TIMEOUT = datetime.timedelta(minutes=5).seconds
+
+    def __init__(self, protocol_util):
+        self._protocol = protocol_util.get_protocol()
+        self.should_run = True
+        self._thread = None
+        self._should_process_events = threading.Event()
+        self._queue = PriorityQueue()
+
+    @staticmethod
+    def get_thread_name():
+        return TelemetryServiceHandler._THREAD_NAME
+
+    def run(self):
+        logger.info("Start Extension Telemetry service.")
+        self.start()
+
+    def is_alive(self):
+        return self._thread is not None and self._thread.is_alive()
+
+    def start(self):
+        self._thread = threading.Thread(target=self._process_telemetry_thread)
+        self._thread.setDaemon(True)
+        self._thread.setName(self.get_thread_name())
+        self._thread.start()
+
+    def stop(self):
+        """
+        Stop server communication and join the thread to main thread.
+        """
+        self.should_run = False
+        if self.is_alive():
+            self.join()
+
+    def join(self):
+        self._queue.join()
+        self._thread.join()
+
+    def stopped(self):
+        return not self.should_run
+
+    def enqueue_event_func(self, event):
+        # Add event to queue and set event
+        self._queue.put(event)
+
+        # Set the event if any enqueue happens (even if already set) to trigger sending those events
+        self._should_process_events.set()
+
+    def _process_telemetry_thread(self):
+        logger.info("Successfully started the {0} thread".format(self.get_thread_name()))
+        try:
+            # On demand wait, start processing as soon as there is any data available in the queue
+            # In worst case, also keep checking every 5 mins to ensure that no data is being missed
```

Comment:
Regardless of the synchronization issue above, enqueue should raise if somebody tries to add an event when the telemetry thread has already been stopped.

Reply:
I agree that if the `self.should_run` property is False, we should raise and not let anyone enqueue events after that but I dont agree on raising when the thread has been stopped. Mainly because the thread can die due to some errors but we have code to restart it. The benefit of adding the Queue in the Handler object was that its independent of the lifecycle of the thread. So as soon as some other thread has an event ready, we should add it to the queue and the telemetry thread can start processing those as soon as its ready to do so. 

Even if the thread dies midway, our queue is not impacted and our data is not lost, which is an improvement from the previous design because incase the `collect_and_send_events` threw error earlier, all that unprocessed data was lost and telemetry was lost.

Reply:
Done